### ISSUE TITLE
[FIX-010] Isolate built trainer configuration from mutable mappings

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#209]
**Category:** Configuration and compatibility · **Priority:** P1 · **Size:** S

**Evidence status:** Reproduced contract defect.

## Outcome and scope

Capture and expose read-only optimizer, scheduler and extra-metric mappings at NNTrainerParams construction. Keep builders mutable and retain runtime loader/callable identity. Fixing only `builder.build()` would leave direct construction and dataclasses.replace vulnerable; the dataclass boundary owns this guarantee.

## Current behavior / evidence

**Context and evidence.** `NNTrainerParams` advertises a frozen configuration, but `NNTrainerParamsBuilder.build()` hands over its live optimizer/scheduler dictionaries. A caller can build configuration A, add an optimizer for configuration B, and silently change A's serialized state and run identity. Direct construction with a caller-owned dictionary has the same weakness. Evidence: [builder](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/trainer/params_builder.py#L137), [params boundary](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/trainer/params.py#L69). Isolated probe confirmed A's keys become `['first', 'second']` after modifying the builder.

## Repository map

- [`src/nnx/trainer/params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/trainer/params.py#L69) — Construction, key validation and state serialization boundary.
- [`src/nnx/trainer/params_builder.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/trainer/params_builder.py#L138) — Builder passes live mappings to the frozen dataclass.
- [`src/nnx/nn/params/nn_evaluation_data_point.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_evaluation_data_point.py#L11) — Existing small Mapping pattern; values here need not be floats or sorted at runtime.
- [`tests/test_nn_trainer_params_builder.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_nn_trainer_params_builder.py#L30) — Mutable builder and omission compatibility tests.
- [`tests/test_trainer_params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_trainer_params.py#L41) — Direct construction and state/from_state coverage.
- [`src/nnx/nn/params/nn_run.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_run.py#L236) — Run identity consumes serialized trainer state; runtime insertion order must remain intact.

## Implementation approach

1. Snapshot each supplied mapping with a new insertion-ordered dict inside `NNTrainerParams.__post_init__`, before validating optimizer/scheduler keys. Preserve `extra_metrics=None` versus an explicitly supplied empty mapping; capture nonempty metric registrations without wrapping or copying callables.
2. Expose a small owned read-only Mapping supporting normal iteration, lookup, equality and the supported copy/pickle paths. Reuse the project’s owned-Mapping pattern where appropriate, but do not use a bare MappingProxyType without proving deepcopy/pickle behavior. Values are existing immutable params or intentional shared runtime callables; do not deepcopy DataLoaders or live modules.
3. Let builder setters retain in-place/last-call-wins behavior. Each `build()` captures a new value via the dataclass boundary; direct constructor, `from_state` and `dataclasses.replace` must receive the same protection. Assignment/deletion through the built mapping should raise TypeError.
4. Preserve runtime optimizer insertion order. `state()` deliberately sorts keys and omits empty/default fields: keep byte-equivalent serialized content and historical run IDs. Document what is copied (configuration containers) and what is shared (loader/callable objects), without adding builder branching APIs from FEAT-041.

## Acceptance criteria

- [ ] Mutating or reusing a builder never changes an already-built configuration's mappings or `state()` output.
- [ ] Mutating input dictionaries after direct construction also leaves the configuration unchanged; normal item assignment/deletion through the built mapping is rejected.
- [ ] Optimizer/scheduler key validation happens against the captured snapshot; metric callables remain callable.
- [ ] `state()`/`from_state()`, `dataclasses.replace`, documented copying, and legacy default run-ID fixtures continue to work.
- [ ] Tests cover empty/nonempty schedulers and extra metrics as well as optimizers; no attempt is made to freeze live DataLoader internals.

## Validation and completion

Runnable existing-API isolation fixture:

```python
from dataclasses import replace
from nnx import NNOptimParams, NNTrainerParams

opt = NNOptimParams.builder().adam(max_lr=1e-3).build()
builder = NNTrainerParams.builder().n_epochs(2).optimizer("first", opt)
first = builder.build()
saved = first.state()
builder.optimizer("second", opt)
assert set(first.optims) == {"first"}
assert first.state() == saved
assert replace(first, n_epochs=3).optims == first.optims
source = {"first": opt}
direct = NNTrainerParams(n_epochs=2, optims=source)
source.clear()
assert set(direct.optims) == {"first"}
```

Proposed cases: `test_built_trainer_params_snapshot_mappings`, `test_direct_trainer_params_reject_mapping_mutation`, and `test_trainer_mapping_copy_and_roundtrip`. Include nonempty and empty schedulers, metric callable identity/callability, mutation through the original dict and built mapping, shallow/deep copy, replace, and pickle round-trip for pickleable inputs. Use module-level callable fixtures for pickle tests; lambdas need not become pickleable. Compare serialized snapshots and existing ID fixtures, and verify loader identity survives construction unchanged.

From the repository root, use `uv sync --frozen --all-extras` for the CI-equivalent environment. Existing focused coverage (files already exist):

```bash
uv run --frozen pytest tests/test_nn_trainer_params_builder.py tests/test_trainer_params.py tests/test_params_round_trip.py
```

Add the proposed regression cases above to those existing test files, then rerun the same command. Finish with:

```bash
uv run --frozen ruff check src/ tests/ examples/ scripts/
uv run --frozen ruff format --check src/ tests/ examples/ scripts/
uv run --frozen pyright --warnings
```

No new dependency is needed for the implementation described. Preserve Python 3.10 compatibility and default-field omission in existing serialization. These commands are completion requirements for implementation; this specification review does not claim they were run or passed.

## Cascading changes and affected consumers

- **Runtime versus descriptor snapshot.** [NNTrainerParamsBuilder.build](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/trainer/params_builder.py) and direct construction must hand an owned immutable mapping snapshot to [Trainer.train](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/trainer/trainer.py). Verify the actual created optimizer/scheduler names and parameter-group membership after callers mutate original dictionaries or reuse a builder; checking `state()` alone can miss runtime drift. Preserve deterministic sorted primary-optimizer selection for callbacks and existing optimizer construction/step order. Runtime optimizer objects, loader objects and metric callables retain their existing identities; only configuration containers become immutable.

- **Run identity and round trips.** [NNRun.state/id](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_run.py) caches a descriptor hash, so a later builder mutation must not create a run whose recorded descriptor differs from the optimizers it executes. Exercise YAML save/load, `from_state`, `dataclasses.replace`, and `with_train_loader`/`with_val_loader` through a real small Trainer run. Emit ordinary serialized dictionaries with the current sorted keys and default omissions, not an implementation-specific mapping wrapper. Runtime-only extra metric callables must stay out of persisted state and remain callable in the captured configuration; any supported copy/pickle behavior must not imply arbitrary closures are pickleable.

- **Multi-optimizer teaching.** [Example 09](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/09_gan_with_trainer.py) builds a two-optimizer configuration directly and currently runs a much larger teaching workload. Add a tiny builder-reuse helper: capture a G/D config, modify the builder for a second config, then execute only the first for two minibatches. Assert that callbacks and persisted descriptor still refer to its original names and that both scoped optimizers update their intended parameters. The docs should distinguish intentionally reusable mutable builders from immutable built configurations, rather than merely calling the outer dataclass frozen.

### Samples and executable coverage

- [`examples/09_gan_with_trainer.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/09_gan_with_trainer.py): Add a tiny builder-reuse helper that captures G/D config, mutates the builder for another config, then runs the captured config for two minibatches and checks names, scoped updates and persisted state. **Verify:** Proposed execution node: uv run --frozen pytest tests/test_examples_smoke.py -k trainer_builder_snapshot. CPU, offline, temporary working directory; call the bounded helper rather than the full training main. Retain import-without-main coverage and add the callable to explicit execution coverage; import success alone is insufficient.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`docs/concepts.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/concepts.md) | `generated/site/Concepts.md` | `generated/wiki/Concepts.md` | Explain snapshot ownership for Trainer optims/schedulers/extra_metrics, builder reuse, runtime-only callable/loader identity, and stable sorted/default-omitted serialization. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] A real Trainer run built before caller/builder mapping mutation creates only the captured optimizer/scheduler set, preserves scoped parameter ownership and exposes the original deterministic primary optimizer to callbacks.
- [ ] The executed optimizer configuration matches its cached NNRun identity and saved YAML descriptor after later builder reuse, and reloading reconstructs equivalent immutable mappings with legacy omissions unchanged.
- [ ] with_train_loader, with_val_loader and dataclasses.replace preserve the captured mapping contents while retaining intended runtime loader/callable identities and excluding extra metric callables from serialized state.
- [ ] The bounded Example 09 builder-reuse helper executes the original two-optimizer config after builder mutation and checks both scoped updates and saved descriptor contents.
- [ ] The issue-specific changes in `docs/concepts.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** Propagation belongs to this fix and existing APIs. Related tickets coordinate shared contracts/files; they are not prerequisites for the bounded repaired path.

Additional nonblocking consumers/coordination: [FEAT-005](https://github.com/thekaveh/NNx/issues/215), [FEAT-012](https://github.com/thekaveh/NNx/issues/221), [FEAT-041](https://github.com/thekaveh/NNx/issues/258).

## Delivery order and dependencies

- **Execution order:** 10 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.
- **Related / optional integration:** [FEAT-012](https://github.com/thekaveh/NNx/issues/221). These links do not block this ticket.

**Dependency rationale.** Independent repair of current behavior. No dependency; prerequisite where FEAT-012 reuses these configurations.

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved.

**Root cause:** `NNTrainerParams` kept the caller's live `optims` / `schedulers` / `extra_metrics` dicts by reference (the builder handed over its own working dicts), so reusing a builder or mutating a source dict silently changed an already-built configuration, its `state()` and the run identity derived from it.

**What changed**
- `NNTrainerParams.__post_init__` snapshots the three mappings into `_FrozenMapping` (insertion-ordered, read-only `Mapping`) before key validation; covers the direct constructor, `builder().build()`, `from_state`, `dataclasses.replace`, `with_train_loader` / `with_val_loader`.
- Item assignment / deletion on the built mappings raise `TypeError`; values keep identity (immutable params dataclasses, runtime metric callables, loaders — no deep copies). `copy` / `deepcopy` / `pickle` rebuild equivalent snapshots. `state()` output and run IDs are byte-identical; builders stay mutable and last-call-wins.
- Example 09: helpers hoisted to module level plus a bounded `trainer_builder_snapshot()` (executed by `tests/test_examples_smoke.py -k trainer_builder_snapshot`) that runs a captured G/D config after builder reuse and checks optimizer names, scoped parameter ownership/updates, the sorted-first primary seen by callbacks, and the persisted descriptor. Docs: `concepts.md` §8, API reference, examples catalog, CHANGELOG.

**PRs:** #269 (develop, merged), #270 (main, merged).

**Checks run (both branches):**
- `uv run --frozen pytest -q` → 1593 passed, 1 skipped, 2 deselected (the two libcairo-dependent `test_docs_projection` tests — pre-existing local-env failures, unrelated; CI runs them and passed)
- `ruff check` → All checks passed; `ruff format --check` → all formatted; `pyright --warnings` → 0 errors, 0 warnings
- `build_api_reference` / `build_docs` / `mkdocs build --strict` / `build_wiki --check` / docs+wiki projection tests → pass
- CI: lint-and-test on Python 3.10–3.14, pip-audit, CodeQL, docs build all green on #269 and #270.

**Not verifiable locally:** nothing hardware-specific; live Pages/wiki publication is left to the `docs.yml` main-push job.
